File: core/player/prop/explosion.js
import { assets } from "../../assets/main.js";
import { randomNum } from "../../utils/random.js";
import { ctx } from "../../world/canvas.js";
import { drawWrapped, wrap } from "../../world/utils.js";
import { world } from "../../world/world.js";

export default class Explosion {
    constructor(x, y, intensity = 100) {
        this.particles = Array.from(new Array(Math.ceil(intensity/10))).map((p) => ({ x, y, vx: randomNum(-intensity, intensity) / 10, vy: randomNum(-intensity, intensity) / 10, radius: randomNum(0.1, 2) }));
        this.life = 50;
        this.img = assets?.images?.explosionflame
    }

    render() {
        ctx.fillStyle = "red";
        for (const p of this.particles) {
            drawWrapped({
                fn: () => {
                    if (this.img) {
                        ctx.drawImage(this.img, -5, -5, 10, 10);
                    }
                    else {
                        ctx.beginPath();
                        ctx.arc(0, 0, p.radius, 0, Math.PI * 2);
                        ctx.fill();
                    }
                }, x: p.x, y: p.y
            })

            p.x = wrap(p.x + p.vx, world.width);
            p.y = wrap(p.y + p.vy, world.height)
        }

        if (!(--this.life)) this.destroy();
    }

    destroy() {
        const index = explosions.indexOf(this);
        if (index > -1) {
            explosions.splice(index, 1);
        }
    }
}

export const explosions = [];

File: void_front/game/core/player/prop/explosion.py
from __future__ import annotations

from dataclasses import dataclass, field

from game.core.assets.main import assets
import numpy as np
from game.core.world.canvas import blit_image, draw_circle
from game.core.world.utils import drawWrapped, wrap
from game.core.world.world import world


explosions: list["Explosion"] = []


@dataclass
class Explosion:
    x: float
    y: float
    intensity: int = 100
    life: int = 50
    particles: list[dict] = field(default_factory=list)

    def __post_init__(self):
        if not self.particles:
            count = max(int(self.intensity / 10), 1)
            self.particles = [
                {
                    "x": self.x,
                    "y": self.y,
                    "vx": np.random.uniform(-self.intensity, self.intensity) / 10,
                    "vy": np.random.uniform(-self.intensity, self.intensity) / 10,
                    "radius": np.random.uniform(0.1, 2),
                }
                for _ in range(count)
            ]
        self.img = getattr(getattr(assets, "images", None), "explosionflame", None)

    def render(self):
        for particle in self.particles:
            if self.img is not None:
                blit_image(self.img, (particle["x"], particle["y"]), size=(10, 10), angle=0, alpha=255)
            else:
                draw_circle((particle["x"], particle["y"]), particle["radius"], color="red", alpha=255)
            particle["x"] = wrap(particle["x"] + particle["vx"], world.width)
            particle["y"] = wrap(particle["y"] + particle["vy"], world.height)

        self.life -= 1
        if self.life <= 0:
            self.destroy()

    def destroy(self):
        if self in explosions:
            explosions.remove(self)
